### PR TITLE
Add JUnit 5 `ScorerExtension` for AI model evaluation and associated library

### PR DESCRIPTION
**Pull request by cescoffier**
- Implement `ScorerExtension` to inject and manage Scorer instances in tests.
- Support field and parameter injection for Scorer using `@ScorerConfiguration`.
- Add support for parameter injection of samples via `@SampleLocation` annotation.
- Provide built-in evaluation strategies:
  - `SemanticSimilarityStrategy` (cosine similarity-based evaluation).
  - `AiJudgeStrategy` (AI-powered evaluation with customizable prompts).
- Add tests for ScorerExtension:
  - Validate field and parameter injection of Scorer.
  - Test sample injection from YAML files.
  - Verify evaluation strategies and reporting.
- Document ScorerExtension:
  - Explain concepts: Scorer, Samples, Evaluation Strategies, Reports.
  - Usage examples for field/parameter injection and evaluation.
  - Guide for using built-in strategies and creating custom strategies.

### CODE REVIEW

File: docs/modules/ROOT/pages/testing.adoc [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+import io.quarkiverse.langchain4j.scorer.junit5.AiScorer;
+
+@AiScorer
+public class MyScorerTests {
+
+    // Test cases go here
+}
+----
+
+This Junit5 extension can be combined with `@QuarkusTest` to test Quarkus applications:
```
Comment (gsmet) [inline docs/modules/ROOT/pages/testing.adoc:51]:
```suggestion
This JUnit 5 extension can be combined with `@QuarkusTest` to test Quarkus applications:
```

File: docs/modules/ROOT/pages/testing.adoc [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+
+    // Test cases go here
+}
+----
+
+== Concepts
+
+=== Scorer
+
+The `Scorer` (`io.quarkiverse.langchain4j.testing.scorer.Scorer`) is a utility that evaluates a _set of samples_ (represented by `io.quarkiverse.langchain4j.testing.scorer.Samples`) against a function (parts of the application) and a set of evaluation strategies.
```
Comment (gsmet) [inline docs/modules/ROOT/pages/testing.adoc:72]:
```suggestion
The `Scorer` (`io.quarkiverse.langchain4j.testing.scorer.Scorer`) is a utility that evaluates a _set of samples_ (represented by `io.quarkiverse.langchain4j.testing.scorer.Samples`) against a function (part of the application) and a set of evaluation strategies.
```

File: docs/modules/ROOT/pages/testing.adoc [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+    assertThat(report.score()).isGreaterThanOrEqualTo(70); // Assert the score
+}
+----
+
+=== Samples
+
+A `Sample` (`io.quarkiverse.langchain4j.testing.scorer.EvaluationSample`) represents a single input-output test case.
+It includes:
+- a name: the name of the sample,
+- the inputs: The input data for the test,
```
Comment (gsmet) [inline docs/modules/ROOT/pages/testing.adoc:97]: I would drop the capitalization here and for the items below or be consistent for the first item (but I recommend the former).

File: docs/modules/ROOT/pages/testing.adoc [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+
+In general, tests using the `Scorer` follow this pattern:
+
+[source, java]
+----
+ @Test
+void testAiService(@ScorerConfiguration(concurrency = 5) Scorer scorer, // The scorer instance, with concurrency set to 5
+                   @SampleLocation("src/test/resources/samples.yaml") Samples<String> samples) { // The samples loaded from a YAML file
+    // `aiServiceEvaluation` is a function that takes the sample inputs and returns an output - this is the evaluated output
+    EvaluationReport report = scorer.evaluate(samples, aiServiceEvaluation,
```
Comment (gsmet) [inline docs/modules/ROOT/pages/testing.adoc:86]: I think having the type of `aiServiceEvaluation` would help to understand things.

File: docs/modules/ROOT/pages/testing.adoc [hunk 5]
```diff
@@ -0,0 +1,624 @@
...
+}
+----
+
+=== Evaluation Report
+
+The `EvaluationReport` aggregates the results of all evaluations. It provides:
+
+- A global score (percentage of passed evaluations).
+- Scores per tag.
+- the possibility to dump the report as Markdown.
```
Comment (gsmet) [inline docs/modules/ROOT/pages/testing.adoc:170]: Inconsistent capitalization.

**Review by gsmet — COMMENTED** [review]
I read the doc to better understand this thing and I noticed a few tiny things (but feel free to ignore, they are tiny tiny).

File: docs/modules/ROOT/pages/testing.adoc (hunk 1 above)
Comment (cescoffier) [reply docs/modules/ROOT/pages/testing.adoc]: Fixed.

File: docs/modules/ROOT/pages/testing.adoc (hunk 3 above)
Comment (cescoffier) [reply docs/modules/ROOT/pages/testing.adoc]: Fixed.

File: testing/scorer/scorer-strategies/ai-judge/src/main/java/io/quarkiverse/langchain4j/testing/scorer/judge/AiJudgeStrategy.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+     * Evaluate the output of a model.
+     *
+     * @param sample the sample to evaluate.
+     * @param output the output of the model.
+     * @return {@code true} if the output is correct, {@code false} otherwise.
+     */
+    @Override
+    public boolean evaluate(EvaluationSample<String> sample, String output) {
+        String expectedOutput = sample.expectedOutput();
+        String prompt = this.prompt
```
Comment (geoand) [inline testing/scorer/scorer-strategies/ai-judge/src/main/java/io/quarkiverse/langchain4j/testing/scorer/judge/AiJudgeStrategy.java:55]: I could see this being an actual template in the future, but no reason to overcomplicate for now

File: docs/modules/ROOT/pages/testing.adoc (hunk 4 above)
Comment (cescoffier) [reply docs/modules/ROOT/pages/testing.adoc]: Done.

File: docs/modules/ROOT/pages/testing.adoc (hunk 2 above)
Comment (cescoffier) [reply docs/modules/ROOT/pages/testing.adoc]: Fixed.

File: testing/scorer/scorer-junit5/src/main/java/io/quarkiverse/langchain4j/scorer/junit5/AiScorer.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.TYPE })
+@ExtendWith(ScorerExtension.class)
+public @interface AiScorer {
```
Comment (geoand) [inline testing/scorer/scorer-junit5/src/main/java/io/quarkiverse/langchain4j/scorer/junit5/AiScorer.java:13]: We'll need some Javadoc for this

File: docs/modules/ROOT/pages/testing.adoc (hunk 5 above)
Comment (cescoffier) [reply docs/modules/ROOT/pages/testing.adoc:180]: Fixed.

File: testing/scorer/scorer-junit5/src/main/java/io/quarkiverse/langchain4j/scorer/junit5/SampleLocation.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package io.quarkiverse.langchain4j.scorer.junit5;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.PARAMETER)
+public @interface SampleLocation {
```
Comment (geoand) [inline testing/scorer/scorer-junit5/src/main/java/io/quarkiverse/langchain4j/scorer/junit5/SampleLocation.java:10]: Same

File: testing/scorer/scorer-junit5/src/main/java/io/quarkiverse/langchain4j/scorer/junit5/ScorerConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+import static java.lang.annotation.ElementType.*;
+import static java.lang.annotation.RetentionPolicy.*;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+@Retention(RUNTIME)
+@Target({ FIELD, PARAMETER })
+public @interface ScorerConfiguration {
```
Comment (geoand) [inline testing/scorer/scorer-junit5/src/main/java/io/quarkiverse/langchain4j/scorer/junit5/ScorerConfiguration.java:11]: And here

**Review by geoand — COMMENTED** [review]
Very neat, this was sorely needed!

I added a few small comments

File: testing/scorer/scorer-core/src/main/java/io/quarkiverse/langchain4j/testing/scorer/EvaluationStrategy.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+public interface EvaluationStrategy<T> {
+
+    /**
+     * Evaluate the output of a model.
+     *
+     * @param sample the sample to evaluate.
+     * @param output the output of the model.
+     * @return {@code true} if the output is correct, {@code false} otherwise.
+     */
+    boolean evaluate(EvaluationSample<T> sample, T output);
```
[the remark this answers is not in the document]
Comment (cescoffier) [reply testing/scorer/scorer-core/src/main/java/io/quarkiverse/langchain4j/testing/scorer/EvaluationStrategy.java:17]: For now, yes. 

Strategies using:

- multiple booleans by characteristics are unusable
- sub-scoring leads to very complex math to compute the global score, which at the end, cannot be reasoned about

Not that this is for a single strategy and called for every sample.

File: testing/scorer/scorer-core/src/main/java/io/quarkiverse/langchain4j/testing/scorer/Input.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
+package io.quarkiverse.langchain4j.testing.scorer;
+
+import org.eclipse.microprofile.config.spi.Converter;
+
+import io.smallrye.config.Converters;
+
+/**
+ * Represents a parameter passed to the function to evaluate.
+ */
+public interface Input {
```
[the remark this answers is not in the document]
Comment (cescoffier) [reply testing/scorer/scorer-core/src/main/java/io/quarkiverse/langchain4j/testing/scorer/Input.java:10]: Does `Parameter` look better?

File: testing/scorer/scorer-strategies/ai-judge/src/main/java/io/quarkiverse/langchain4j/testing/scorer/judge/AiJudgeStrategy.java (hunk 6 above)
Comment (cescoffier) [reply testing/scorer/scorer-strategies/ai-judge/src/main/java/io/quarkiverse/langchain4j/testing/scorer/judge/AiJudgeStrategy.java:55]: Yes. This strategy needs a bit of tuning.

File: testing/scorer/scorer-core/src/main/java/io/quarkiverse/langchain4j/testing/scorer/Input.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (geoand) [reply testing/scorer/scorer-core/src/main/java/io/quarkiverse/langchain4j/testing/scorer/Input.java:10]: Yeah, that does sound a little better

File: testing/scorer/scorer-junit5/src/main/java/io/quarkiverse/langchain4j/scorer/junit5/AiScorer.java (hunk 7 above)
Comment (cescoffier) [reply testing/scorer/scorer-junit5/src/main/java/io/quarkiverse/langchain4j/scorer/junit5/AiScorer.java:17]: Done.

File: testing/scorer/scorer-junit5/src/main/java/io/quarkiverse/langchain4j/scorer/junit5/SampleLocation.java (hunk 8 above)
Comment (cescoffier) [reply testing/scorer/scorer-junit5/src/main/java/io/quarkiverse/langchain4j/scorer/junit5/SampleLocation.java:15]: Done.

File: testing/scorer/scorer-junit5/src/main/java/io/quarkiverse/langchain4j/scorer/junit5/ScorerConfiguration.java (hunk 9 above)
Comment (cescoffier) [reply testing/scorer/scorer-junit5/src/main/java/io/quarkiverse/langchain4j/scorer/junit5/ScorerConfiguration.java:16]: Done.

File: testing/scorer/scorer-core/src/main/java/io/quarkiverse/langchain4j/testing/scorer/Input.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (cescoffier) [reply testing/scorer/scorer-core/src/main/java/io/quarkiverse/langchain4j/testing/scorer/Input.java]: Fixed

**Review by geoand — APPROVED** [review]
🎉

**Pull request merged**